DeserializeException

From time to time I get the following exception during checkin process of one model:

```
8113966 [pool-1-thread-1] ERROR org.bimserver.longaction.LongAction - [UNKNOWN_DESERIALIZER_ERROR] [firebox.ifc] Error on line 2370:  (null) #6329= IFCMATERIALLAYER(#6330,15.,$,'Layer',$,$,$);
org.bimserver.shared.exceptions.UserException: [firebox.ifc] Error on line 2370:  (null) #6329= IFCMATERIALLAYER(#6330,15.,$,'Layer',$,$,$);
	at org.bimserver.database.actions.StreamingCheckinDatabaseAction.execute(StreamingCheckinDatabaseAction.java:412)
	at org.bimserver.database.actions.StreamingCheckinDatabaseAction.execute(StreamingCheckinDatabaseAction.java:84)
	at org.bimserver.database.DatabaseSession.executeAndCommitAction(DatabaseSession.java:775)
	at org.bimserver.database.DatabaseSession.executeAndCommitAction(DatabaseSession.java:752)
	at org.bimserver.longaction.LongStreamingCheckinAction.execute(LongStreamingCheckinAction.java:73)
	at org.bimserver.longaction.LongActionManager$1.run(LongActionManager.java:55)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.bimserver.plugins.deserializers.DeserializeException: Error on line 2370:  (null) #6329= IFCMATERIALLAYER(#6330,15.,$,'Layer',$,$,$);
	at org.bimserver.ifc.step.deserializer.IfcStepStreamingDeserializer.read(IfcStepStreamingDeserializer.java:205)
	at org.bimserver.ifc.step.deserializer.IfcStepStreamingDeserializer.read(IfcStepStreamingDeserializer.java:175)
	at org.bimserver.database.actions.StreamingCheckinDatabaseAction.execute(StreamingCheckinDatabaseAction.java:196)
	... 10 common frames omitted
Caused by: java.nio.BufferOverflowException: null
	at java.base/java.nio.Buffer.nextPutIndex(Unknown Source)
	at java.base/java.nio.HeapByteBuffer.putLong(Unknown Source)
	at org.bimserver.shared.ByteBufferVirtualObject.<init>(ByteBufferVirtualObject.java:62)
	at org.bimserver.ifc.step.deserializer.IfcStepStreamingDeserializer.newVirtualObject(IfcStepStreamingDeserializer.java:304)
	at org.bimserver.ifc.step.deserializer.IfcStepStreamingDeserializer.processRecord(IfcStepStreamingDeserializer.java:333)
	at org.bimserver.ifc.step.deserializer.IfcStepStreamingDeserializer.processLine(IfcStepStreamingDeserializer.java:252)
	at org.bimserver.ifc.step.deserializer.IfcStepStreamingDeserializer.read(IfcStepStreamingDeserializer.java:193)
	... 12 common frames omitted
```

The strange thing is that all IFC viewers correctly read the model. And BIMServer itself after some activities such as uploading other models correctly uploads this model. But sometimes it crushes with this error. Other instances of BIMServer checkin this model also without errors. But at one time something happens and it throws error.

--- a/Shared/src/org/bimserver/shared/ByteBufferVirtualObject.java
+++ b/Shared/src/org/bimserver/shared/ByteBufferVirtualObject.java
@@ -50,14 +50,13 @@ public class ByteBufferVirtualObject extends AbstractByteBufferVirtualObject imp
 	private Map<Integer, ByteBufferList> referencedBuffers;
 	
 	public ByteBufferVirtualObject(QueryContext reusable, EClass eClass, int capacity) {
-		super(Math.max(capacity, 17)); // 17 bytes are always needed
+		super(Math.max(capacity, reusable.getPackageMetaData().getUnsettedLength(eClass) + 16)); // 16 bytes plus the class specific ones are always needed
 		this.reusable = reusable;
 		this.eClass = eClass;
 		this.oid = reusable.getDatabaseInterface().newOid(eClass);
 		this.uuid = reusable.getDatabaseInterface().newUuid();
 
 		int unsettedLength = reusable.getPackageMetaData().getUnsettedLength(eClass);
-		ensureCapacity(buffer.position(), 16 + unsettedLength);
 		buffer.put(new byte[unsettedLength]);
 		buffer.putLong(this.uuid.getMostSignificantBits());
 		buffer.putLong(this.uuid.getLeastSignificantBits());

--- a/Shared/src/org/bimserver/shared/ByteBufferVirtualObject.java
+++ b/Shared/src/org/bimserver/shared/ByteBufferVirtualObject.java
@@ -57,6 +57,7 @@ public class ByteBufferVirtualObject extends AbstractByteBufferVirtualObject imp
 		this.uuid = reusable.getDatabaseInterface().newUuid();
 
 		int unsettedLength = reusable.getPackageMetaData().getUnsettedLength(eClass);
+		ensureCapacity(buffer.position(), 16 + unsettedLength);
 		buffer.put(new byte[unsettedLength]);
 		buffer.putLong(this.uuid.getMostSignificantBits());
 		buffer.putLong(this.uuid.getLeastSignificantBits());

